File: 2025/day07.cpp
// https://adventofcode.com/2025/day/7

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <iostream>
#include <print>
#include <ranges>

using Manifolds = aoc::char_grid<>;
using Input = std::pair<Manifolds, point>;
constexpr let empty = '.';
constexpr let splitter = '^';

fn parse(String const& filename) -> Input {
  auto [manifolds, config] = aoc::read_char_grid(
      filename, {.padding = {}, .start_char = 'S', .end_char = {}});
  return {std::move(manifolds), *config.start_pos};
}

template <bool quantum>
fn solve_case(Input const& input) -> u64 {
  let & [ manifolds, start_pos ] = input;
  auto num_splits = 0;
  auto row_timelines = Vec(manifolds.num_columns(), u64{});
  row_timelines[start_pos.x] = 1;
  for (let row :
       Range{static_cast<usize>(start_pos.y + 1), manifolds.num_rows()}) {
    auto next_timelines = Vec(manifolds.num_columns(), u64{});
    for (let[column, timelines] : row_timelines | std::views::enumerate) {
      let value = manifolds.at(row, column);
      switch (value) {
        case empty:
          next_timelines[column] += timelines;
          break;
        case splitter: {
          num_splits += static_cast<u64>(timelines > 0);
          next_timelines[column - 1] += timelines;
          next_timelines[column + 1] += timelines;
        } break;
        default:
          AOC_UNREACHABLE("Invalid value in the manifold");
      };
    }
    row_timelines = std::move(next_timelines);
  }
  if constexpr (!quantum) {
    return num_splits;
  } else {
    return aoc::ranges::accumulate(row_timelines, u64{});
  }
}

int main() {
  std::println("Part 1");
  let example = parse("day07.example");
  AOC_EXPECT_RESULT(21, solve_case<false>(example));
  let example2 = parse("day07.example2");
  AOC_EXPECT_RESULT(16, solve_case<false>(example2));
  let input = parse("day07.input");
  AOC_EXPECT_RESULT(1560, solve_case<false>(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(40, solve_case<true>(example));
  AOC_EXPECT_RESULT(26, solve_case<true>(example2));
  AOC_EXPECT_RESULT(25592971184998, solve_case<true>(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2025/src/day07.rs
type Manifolds = aoc::grid::Grid<char>;
type Upos = aoc::grid::Upos;
const EMPTY: char = '.';
const SPLITTER: char = '^';

fn parse(filename: &str) -> (Manifolds, Upos) {
    let (manifolds, config) = Manifolds::from_file_config(
        filename,
        aoc::grid::ConfigInput {
            padding: None,
            start_char: Some('S'),
            end_char: None,
        },
    );
    (manifolds, config.start_pos.unwrap())
}

fn solve_case<const QUANTUM: bool>((manifolds, start_pos): &(Manifolds, Upos)) -> u64 {
    let mut num_splits = 0;
    let mut row_timelines = vec![0; manifolds.num_columns];
    row_timelines[start_pos.x] = 1;
    for row in (start_pos.y + 1)..(manifolds.num_rows) {
        let mut next_timelines = vec![0; manifolds.num_columns];
        for (column, &timelines) in row_timelines.iter().enumerate() {
            let value = *manifolds.get(row, column);
            match value {
                EMPTY => {
                    next_timelines[column] += timelines;
                }
                SPLITTER => {
                    num_splits += (timelines > 0) as u64;
                    next_timelines[column - 1] += timelines;
                    next_timelines[column + 1] += timelines;
                }
                _ => unreachable!("Invalid value in the manifold"),
            };
        }
        row_timelines = next_timelines;
    }
    return if (!QUANTUM) {
        num_splits
    } else {
        row_timelines.iter().sum()
    };
}

fn main() {
    println!("Part 1");
    let example = parse("day07.example");
    aoc::expect_result!(21, solve_case::<false>(&example));
    let example2 = parse("day07.example2");
    aoc::expect_result!(16, solve_case::<false>(&example2));
    let input = parse("day07.input");
    aoc::expect_result!(1560, solve_case::<false>(&input));

    println!("Part 2");
    aoc::expect_result!(40, solve_case::<true>(&example));
    aoc::expect_result!(26, solve_case::<true>(&example2));
    aoc::expect_result!(25592971184998, solve_case::<true>(&input));
}
